### PR TITLE
High level publishers subscribers

### PR DESCRIPTION
**Pull request by pdeziel**
This PR introduces high level publishers and subscribers that provide a more object oriented interface for publishing and subscribing. These can be used with inheritance to implement specific interactions with other APIs or simply as a way to avoid writing async code.

I have made the following changes:

1. Added Publisher and Subscriber classes
2. Added tests

### CHECKLIST

- [x] _Is the commit message formatted correctly?_

- [x] _Do all of your functions and methods have docstrings?_
- [x] _Have you added/updated unit tests where appropriate?_
- [x] _Have you run the unit tests using `pytest`?_
- [x] _Is your code style correct (are you using PEP8, pyflakes)?_

### CODE REVIEW

File: pyensign/publisher.py
Comment (pdeziel) [inline pyensign/publisher.py:1]: So the standard use would be something like:

```python
class WeatherPublisher(Publisher):
    def poll_weather(self):
        while True:
            # Call the weather API and create some events
            ...
            sleep(60)
    
    def run_forever(self):
        self.run(self.poll_weather())

publisher = WeatherPublisher("weather")
publisher.run_forever()
```

File: tests/pyensign/test_subscriber.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    @pytest.mark.parametrize(
+        "topics, kwargs, exception",
+        [
+            ("topic", {"client_id": "id"}, ValueError),
+            ("", {"client_id": "id", "client_secret": "secret"}, ValueError),
+        ],
+    )
+    def test_bad_params(self, topics, kwargs, exception):
+        with pytest.raises(exception):
+            Subscriber(*topics, **kwargs)
```
Comment (rebeccabilbro) [inline tests/pyensign/test_subscriber.py:22]: Can you add a note here to explain what about these params are bad? I'm not sure I understand why this should not work

File: tests/pyensign/test_publisher.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        "topic, kwargs, exception",
+        [
+            ("topic", {"client_id": "id"}, ValueError),
+            (None, {"client_id": "id", "client_secret": "secret"}, ValueError),
+            ("", {"client_id": "id", "client_secret": "secret"}, ValueError),
+        ],
+    )
+    def test_bad_params(self, topic, kwargs, exception):
+        with pytest.raises(exception):
+            Publisher(topic, **kwargs)
```
Comment (rebeccabilbro) [inline tests/pyensign/test_publisher.py:31]: What exactly is wrong with the params that should cause this to fail?

File: pyensign/ensign.py [hunk 3]
```diff
@@ -178,7 +178,9 @@ async def subscribe(
         self, *topics, on_event=ack_event, query="", consumer_group=None
     ):
         """
-        Subscribe to events from the Ensign server.
+        Subscribe to events from the Ensign server. This method returns immediately and
+        does not wait for events to be consumed from the topics. To implement a blocking
+        subscriber, callers can use `await asyncio.Future()`.
```
Comment (rebeccabilbro) [inline pyensign/ensign.py:183]: Might need to add just a bit more user-oriented explanation here - does "returns immediately" mean that it essentially returns a cursor that users can use in subsequent calls to read events off the topic?

**Review by rebeccabilbro — APPROVED** [review]
Looking good @pdeziel -- I have a couple of questions for your consideration to help me wrap my mind around how best to coach users to leverage this new interface. Otherwise should be good to go!

File: pyensign/publisher.py
Comment (pdeziel) [reply pyensign/publisher.py:1]: Yeah, that's a good idea.

File: pyensign/publisher.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    """
+
+    def __init__(self, topic, **kwargs):
+        """
+        Create a Publisher to publish events to a topic. Once initialized, call run()
+        to publish events to the topic.
+
+        Parameters
+        ----------
+        topic : str
```
[the remark this answers is not in the document]
Comment (pdeziel) [reply pyensign/publisher.py:20]: Absolutely, I'll clarify that.

File: tests/pyensign/test_publisher.py (hunk 2 above)
Comment (pdeziel) [reply tests/pyensign/test_publisher.py:31]: The first case is that no client secret is provided, the second and third cases are that no topic is provided.

File: tests/pyensign/test_subscriber.py (hunk 1 above)
Comment (pdeziel) [reply tests/pyensign/test_subscriber.py:22]: Sure, the user has to pass in a topic and credentials for their API key (or have credentials in the environment, but the test is assuming no environment).

**Pull request merged**